### PR TITLE
fix: unnecessary_filter_map return of Some arg

### LINKED ISSUE #12556: Suggest removing `filter_map` for `Iterator::filter_map(|x| Some(x))`
### Summary

When we encounter a `filter_map` on what is effectively an identity function followed by `Some`, we shouldn't recommend `map(identity)`, but we should instead recommend completely removing `filter_map`.

Many thanks to @Centri3 for finding this and #12501 .

### Reproducer

I tried this code:

```
fn main() {
    let _= vec![Some(10), None].into_iter().filter_map(|x| Some(x));
}
```

What I saw: 
```
    Checking playground v0.0.1 (/playground)
warning: this `.filter_map` can be written more simply using `.map`
 --> src/main.rs:4:12
  |
4 |     let _= vec![Some(10), None].into_iter().filter_map(|x| Some(x));
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_filter_map
  = note: `#[warn(clippy::unnecessary_filter_map)]` on by default

warning: redundant closure
 --> src/main.rs:4:56
  |
4 |     let _= vec![Some(10), None].into_iter().filter_map(|x| Some(x));
  |                                                        ^^^^^^^^^^^ help: replace the closure with the function itself: `Some`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
  = note: `#[warn(clippy::redundant_closure)]` on by default

warning: `playground` (bin "playground") generated 2 warnings (run `cargo clippy --fix --bin "playground"` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.26s
```

What I expected to see:

```
    Checking playground v0.0.1 (/playground)
warning: this `.filter_map` for `|x| Some(x)` can be more succinctly written as
 --> src/main.rs:4:12
  |
4 |     let _= vec![Some(10), None].into_iter();
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_filter_map
  = note: `#[warn(clippy::unnecessary_filter_map)]` on by default
```

This is a follow up from #12501's discussion.

### Version

```text
Latest Nightly on playground
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by omer-shtivi**
fixes #12556
changelog: [`unnecessary_filter_map`]: filter map return arg as Some

### CODE REVIEW

File: clippy_lints/src/methods/unnecessary_filter_map.rs [hunk 1]
```diff
@@ -36,9 +36,20 @@ pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>, a
...
         let sugg = if !found_filtering {
+            // Check if the closure is .filter_map(|x| Some(x))
+            if name == "filter_map"
+                && let hir::ExprKind::Call(expr, args) = body.value.kind
+            {
+                if is_res_lang_ctor(cx, path_res(cx, expr), OptionSome)
+                    && arg_id.ty_def_id() == args[0].hir_id().ty_def_id()
+                {
+                    if let hir::ExprKind::Path(_) = args[0].kind {
+                        span_lint(cx, UNNECESSARY_FILTER_MAP, arg.span, &format!("{name} is unnecessary"))
```
Comment (J-ZhengLi) [inline clippy_lints/src/methods/unnecessary_filter_map.rs:49]: it just occurs to me that... could this be `span_lint_with_sugg` instead :thinking: , suggesting removing the `filter_map`? 

same thing applies [here](https://github.com/rust-lang/rust-clippy/pull/12766/files#diff-e560e6f117247b6eaa96b371723c123dcefb6eefcf88864d93698a854a0f513aL55-L64), while suggest replacing the method call, with `MaybeIncorrect` applicability for both ofc...

File: tests/ui/unnecessary_filter_map.rs [hunk 2]
```diff
@@ -21,6 +22,8 @@ fn main() {
     //~^ ERROR: this `.filter_map` can be written more simply using `.map`
 
     let _ = (0..4).filter_map(i32::checked_abs);
+
+    let _ = vec![Some(10), None].into_iter().filter_map(|x| Some(x));
```
Comment (belyakov-am) [inline tests/ui/unnecessary_filter_map.rs:26]: The linked issue also mentions passing a path to `Some`, so maybe we should add this as a test too

```rust
let _ = (0..4).filter_map(Some);
```

Comment (llogiq) [reply tests/ui/unnecessary_filter_map.rs:26]: The more tests the merrier 😎

File: clippy_lints/src/methods/unnecessary_filter_map.rs [hunk 3]
```diff
@@ -36,9 +36,20 @@ pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>, a
...
         let in_ty = cx.typeck_results().node_type(body.params[0].hir_id);
         let sugg = if !found_filtering {
+            // Check if the closure is .filter_map(|x| Some(x))
+            if name == "filter_map"
+                && let hir::ExprKind::Call(expr, args) = body.value.kind
+            {
+                if is_res_lang_ctor(cx, path_res(cx, expr), OptionSome)
+                    && arg_id.ty_def_id() == args[0].hir_id().ty_def_id()
+                {
+                    if let hir::ExprKind::Path(_) = args[0].kind {
```
Comment (llogiq) [inline clippy_lints/src/methods/unnecessary_filter_map.rs:48]:
```suggestion
                if is_res_lang_ctor(cx, path_res(cx, expr), OptionSome)
                    && arg_id.ty_def_id() == args[0].hir_id().ty_def_id()
                    && let hir::ExprKind::Path(_) = args[0].kind {
```

let chains for the win!

File: tests/ui/unnecessary_filter_map.rs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+//@no-rustfix
```
Comment (llogiq) [inline tests/ui/unnecessary_filter_map.rs:1]: Is the whole test rustfix-incompatible? If not, could we split our a `-fix` test file?

**Comment by xFrednet** [comment]
Hey @omer-shtivi , this is a ping from triage, since there hasn't been any activity in some time. Are you still planning to continue this implementation?

If you have any questions, you're always welcome to ask them in this PR or on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy).

@rustbot author

**Comment by xFrednet** [comment]
Hey this is triage, I'm closing this due to inactivity. If you want to continue this implementation, you're welcome to create a new PR. Thank you for the time, you already put into this!

Interested parties are welcome to pick this implementation up as well :)

@rustbot label +S-inactive-closed -S-waiting-on-author -S-waiting-on-review

**Pull request closed without merging**